main.py

```diff
@@ -0,0 +1,138 @@
+import random
+import numpy as np
+
+# region Constants
+ELITISM = 1
+SURVIVAL_THRESHOLD = 0.5
+POP_SIZE = 2
+MAX_WEIGHT = 30
+MIN_WEIGHT = -30
+CONNECTION_TYPE = 'Conv'  # could be conv, pool, FC
+# MUTATION_CHANCE = 0.8
+# endregion
+
+
+class Net:
+    def __init__(self, identification, inputs, f=3, k=1):
+        self.identification = identification
+        self.fitness = np.random.randint(10)
+        self.inputs = inputs  # say this is a 5x5 image
+        # self.weights = np.random.uniform(MIN_WEIGHT, MAX_WEIGHT, f**2)
+        self.weights = np.random.randint(-1, 1, f**2)
+        self.weights = np.reshape(self.weights, (f, f))
+        self.f = f
+        self.k = k
+        self.outputs = self.calc_output()
+
+
+    def __str__(self):
+        id_str = f'--Net--\nNet id: {self.identification}\n'
+        fitness_str = f'Fitness: {self.fitness}\n'
+        input_str = f'Input matrix:\n{self.inputs}\n'
+        weight_str = f'Weight kernel:\n{self.weights}\n'
+        output_str = f'Output:\n{self.outputs}\n'
+        return id_str + fitness_str + input_str + weight_str + output_str
+
+    def calc_output(self):
+        in_rows, in_cols = self.inputs.shape
+        f = self.f
+        out_rows = in_rows - f + 1
+        out_cols = in_cols - f + 1
+        outputs = np.zeros(out_rows*out_cols)
+        outputs = np.reshape(outputs, (out_rows, out_cols))
+        kernel = self.weights
+
+        for row in range(out_rows):
+            for col in range(out_cols):
+                filtered = self.inputs[row:row+f, col:col+f] * kernel
+                outputs[row, col] = np.sum(filtered)
+        return outputs
+
+class EvolutionEngine:
+
+    def __init__(self, network_list):
+        self.nets = []
+        self.next_gen = []
+        self.set_nets(network_list)
+
+    def set_nets(self, new_net_list):
+        sorted_list = sorted(new_net_list, key=lambda x: x.fitness)
+        self.nets = sorted_list
+        self.next_gen = []
+
+    def evolve(self):
+        self.elitism()
+
+        num_survivors = int(SURVIVAL_THRESHOLD * len(self.nets))
+        survivor_list = self.nets[0:num_survivors]
+
+        while len(self.next_gen) < POP_SIZE:
+            index = random.randint(num_survivors)
+            mutated_offspring = self.mutate(survivor_list[index])

```

Comment:
What is self.mutate?

Reply:
This will take a net that passes the survival threshold, mutate its hyperparameters and add the offspring to the next generation